clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsMembershipManager.java

```diff
@@ -0,0 +1,1284 @@
+
+        final SortedSet<TopicPartition> partitionsToRelease = topicPartitionsForActiveTasks(activeTasksToRelease);
+        log.debug("Marking lost partitions pending for revocation: {}", partitionsToRelease);
+        subscriptionState.markPendingRevocation(partitionsToRelease);
+
+        return streamsRebalanceEventsProcessor.requestOnAllTasksLostCallbackInvocation();
+    }
+
+    private SortedSet<TopicPartition> partitionsToAssignNotPreviouslyOwned(final SortedSet<TopicPartition> assignedTopicPartitions,
+                                                                           final SortedSet<TopicPartition> ownedTopicPartitions) {
+        SortedSet<TopicPartition> assignedPartitionsNotPreviouslyOwned = new TreeSet<>(TOPIC_PARTITION_COMPARATOR);
+        assignedPartitionsNotPreviouslyOwned.addAll(assignedTopicPartitions);
+        assignedPartitionsNotPreviouslyOwned.removeAll(ownedTopicPartitions);
+        return assignedPartitionsNotPreviouslyOwned;
+    }
+
+    private SortedSet<TopicPartition> topicPartitionsForActiveTasks(final Map<String, SortedSet<Integer>> activeTasks) {
+        final SortedSet<TopicPartition> topicPartitions = new TreeSet<>(TOPIC_PARTITION_COMPARATOR);
+        activeTasks.forEach((subtopologyId, partitionIds) ->
+            Stream.concat(
+                streamsRebalanceData.subtopologies().get(subtopologyId).sourceTopics().stream(),
+                streamsRebalanceData.subtopologies().get(subtopologyId).repartitionSourceTopics().keySet().stream()
+            ).forEach(topic -> {
+                for (final int partitionId : partitionIds) {
+                    topicPartitions.add(new TopicPartition(topic, partitionId));
+                }
+            })
+        );
+        return topicPartitions;
+    }
+
+    private SortedSet<TopicPartition> topicPartitionsForActiveTasks(final SortedSet<StreamsRebalanceData.TaskId> activeTasks) {
+        final SortedSet<TopicPartition> topicPartitions = new TreeSet<>(TOPIC_PARTITION_COMPARATOR);
+        activeTasks.forEach(task ->
+            Stream.concat(
+                streamsRebalanceData.subtopologies().get(task.subtopologyId()).sourceTopics().stream(),
+                streamsRebalanceData.subtopologies().get(task.subtopologyId()).repartitionSourceTopics().keySet().stream()
+            ).forEach(topic -> {
+                topicPartitions.add(new TopicPartition(topic, task.partitionId()));
+            })
+        );
+        return topicPartitions;
+    }
+
+    private void markReconciliationCompleted() {
+        reconciliationInProgress = false;
+        rejoinedWhileReconciliationInProgress = false;
+    }
+
+    private boolean maybeAbortReconciliation() {
+        boolean shouldAbort = state != MemberState.RECONCILING || rejoinedWhileReconciliationInProgress;
+        if (shouldAbort) {
+            String reason = rejoinedWhileReconciliationInProgress ?
+                "the member has re-joined the group" :
```

Comment:
This branch is never tested

Reply:
I added some tests. Admittedly, the number and complexity of the tests tell me that this class needs some refactoring in the future.